In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

# Dán đường dẫn của file vào đây
file_path = '/content/drive/MyDrive/AIO/top 30 main categories.csv'
# Thay '/content/drive/MyDrive/...' bằng đường dẫn thật của bạn

df = pd.read_csv(file_path)

# In 5 dòng đầu tiên của file để kiểm tra
print(df.head())

                                             authors  \
0       Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1  David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2             Jiliang Jing, Qiyuan Pan, Songbai Chen   
3  R. Hole\v{n}\'ak, S. Lohmann, F. Sekula and D....   
4                Szabolcs Horv\'at and Carl D. Modes   

                                               title  \
0  Power-law cosmological solution derived from D...   
1  PyFstat: a Python package for continuous gravi...   
2  Holographic Superconductors with Power-Maxwell...   
3  Simultaneous assessment of energy, charge stat...   
4  Connectedness matters: Construction and exact ...   

                                            abstract  \
0    By studying a tachyon field on the DGP brane...   
1    Gravitational waves in the sensitivity band ...   
2    With the Sturm-Liouville analytical and nume...   
3    We demonstrate simultaneous measurements of ...   
4    We describe a new method for the random s

In [ ]:
# Giả sử DataFrame của bạn đã được đọc vào biến df
# Code để đổi tên các cột
df = df.rename(columns={
    'categories': 'child_labels',
    'main_categories': 'parent_labels'
})

# In ra 5 dòng đầu tiên của DataFrame đã được đổi tên để kiểm tra
print(df.head())

                                             authors  \
0       Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1  David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2             Jiliang Jing, Qiyuan Pan, Songbai Chen   
3  R. Hole\v{n}\'ak, S. Lohmann, F. Sekula and D....   
4                Szabolcs Horv\'at and Carl D. Modes   

                                               title  \
0  Power-law cosmological solution derived from D...   
1  PyFstat: a Python package for continuous gravi...   
2  Holographic Superconductors with Power-Maxwell...   
3  Simultaneous assessment of energy, charge stat...   
4  Connectedness matters: Construction and exact ...   

                                            abstract  \
0    By studying a tachyon field on the DGP brane...   
1    Gravitational waves in the sensitivity band ...   
2    With the Sturm-Liouville analytical and nume...   
3    We demonstrate simultaneous measurements of ...   
4    We describe a new method for the random s

In [ ]:
df = df[['abstract', 'parent_labels', 'child_labels']]

In [ ]:
df['parent_labels'].value_counts()

parent_labels
['hep-th']                  2000
['gr-qc', 'astro-ph']       2000
['hep-th', 'gr-qc']         2000
['physics']                 2000
['quant-ph', 'physics']     2000
['quant-ph']                2000
['quant-ph', 'cond-mat']    2000
['hep-ph', 'nucl-th']       2000
['gr-qc']                   2000
['cond-mat']                2000
['q-bio']                   2000
['gr-qc', 'hep-th']         2000
['nucl-th']                 2000
['hep-ex']                  2000
['hep-ph', 'astro-ph']      2000
['hep-ph', 'hep-ex']        2000
['hep-lat']                 2000
['nlin']                    2000
['econ']                    2000
['eess']                    2000
['cs']                      2000
['math-ph', 'math']         2000
['nucl-ex']                 2000
['hep-th', 'cond-mat']      2000
['hep-ph']                  2000
['math']                    2000
['stat']                    2000
['astro-ph']                2000
['hep-ph', 'hep-th']        2000
['q-fin']                   2000
Name: count, dtype: int64

In [ ]:
df.shape

(60000, 3)

In [ ]:
df.head(3)

abstract          parent_labels  \
0    By studying a tachyon field on the DGP brane...             ['hep-th']   
1    Gravitational waves in the sensitivity band ...  ['gr-qc', 'astro-ph']   
2    With the Sturm-Liouville analytical and nume...    ['hep-th', 'gr-qc']   

        child_labels  
0             hep-th  
1  gr-qc astro-ph.IM  
2       hep-th gr-qc

In [ ]:
type(df['parent_labels'])

pandas.core.series.Series

In [ ]:
df['parent_labels'].head(3)

0               ['hep-th']
1    ['gr-qc', 'astro-ph']
2      ['hep-th', 'gr-qc']
Name: parent_labels, dtype: object

### tai model va luu vao drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
from sentence_transformers import SentenceTransformer
import os

# Định nghĩa đường dẫn tới thư mục đã lưu mô hình trên Drive
# Chú ý: Đây phải là đường dẫn chính xác nơi bạn đã lưu mô hình
model_load_path = '/content/drive/My Drive/Colab_Models/MiniLM-L6-v2'

# Kiểm tra xem đường dẫn có tồn tại không
if not os.path.exists(model_load_path):
    raise FileNotFoundError(f"Lỗi: Không tìm thấy mô hình tại đường dẫn '{model_load_path}'. Vui lòng kiểm tra lại.")

print(f"Đang tải mô hình từ: {model_load_path}")
try:
    # Tải mô hình từ đường dẫn đã lưu trên Drive
    model = SentenceTransformer(model_load_path)
    print("Đã tải mô hình thành công!")

except Exception as e:
    print(f"Lỗi: Có vấn đề khi tải mô hình.")
    print(f"Thông tin chi tiết: {e}")

###embedding văn bản thôi

In [ ]:
import pickle
import numpy as np

In [ ]:
# Tạo embeddings cho cột 'abstract'
print("\nĐang tạo embeddings cho cột 'abstract'...")
abstract_embeddings = model.encode(df['abstract'].tolist(), show_progress_bar=True)
print("Đã tạo xong embeddings cho cột 'abstract'.")
print(f"Kích thước của abstract embeddings: {abstract_embeddings.shape}")


Đang tạo embeddings cho cột 'abstract'...


Batches:   0%|          | 0/938 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [ ]:
# Sử dụng 4 lõi CPU để xử lý song song
num_workers = 4
abstract_embeddings = model.encode(
    df['abstract'].tolist(),
    show_progress_bar=True,
    num_workers=num_workers,
    convert_to_tensor=True
)

Batches:   0%|          | 0/938 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [ ]:
import torch

if torch.cuda.is_available():
    print("GPU có sẵn và đang được sử dụng.")
    print(f"Tên GPU: {torch.cuda.get_device_name(0)}")
else:
    print("GPU không có sẵn. Đang sử dụng CPU.")

GPU có sẵn và đang được sử dụng.
Tên GPU: Tesla T4


In [ ]:
import pandas as pd
import torch
from sentence_transformers import SentenceTransformer
# 1. Kiểm tra và sử dụng GPU nếu có
device = 'cuda' if torch.cuda.is_available() else 'cpu'
if device == 'cuda':
    print("🚀 GPU (CUDA) đã sẵn sàng. Tăng tốc độ tính toán!")
else:
    print("😔 GPU không khả dụng. Đang chạy trên CPU.")

# 2. Tải mô hình và chuyển nó sang GPU
# 'all-MiniLM-L6-v2' là một trong những mô hình nhanh và hiệu quả nhất
model = SentenceTransformer('all-MiniLM-L6-v2', device=device)

# 3. Tạo embeddings với tối ưu hóa
# Tham số num_workers giúp tăng tốc độ nạp dữ liệu vào GPU, tránh "nút cổ chai" CPU
print("\nĐang tạo embeddings cho cột 'abstract'...")
abstract_embeddings = model.encode(
    df['abstract'].tolist(),
    show_progress_bar=True,
    num_workers=2, # Sử dụng 2 luồng CPU để tiền xử lý dữ liệu
    batch_size=256, # Kích thước batch lớn giúp GPU hoạt động hiệu quả hơn
    normalize_embeddings=True # Thêm tham số này để chuẩn hóa tự động
    )
print("Đã tạo xong embeddings cho cột 'abstract'.")
print(f"Kích thước của abstract embeddings: {abstract_embeddings.shape}")

🚀 GPU (CUDA) đã sẵn sàng. Tăng tốc độ tính toán!


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]


Đang tạo embeddings cho cột 'abstract'...


Batches:   0%|          | 0/235 [00:00<?, ?it/s]

Đã tạo xong embeddings cho cột 'abstract'.
Kích thước của abstract embeddings: (60000, 384)


In [ ]:
abstract_embeddings.shape

(60000, 384)

embedding unique label nhãn cha lưu thành

In [ ]:
import ast
import pandas as pd # Giả sử bạn đang dùng pandas

# Hàm để chuyển đổi an toàn, chỉ áp dụng literal_eval nếu là chuỗi
def safe_literal_eval(data):
    if isinstance(data, str):
        try:
            return ast.literal_eval(data)
        except (ValueError, SyntaxError):
            # Trả về một danh sách rỗng hoặc xử lý lỗi nếu chuỗi không hợp lệ
            return []
    # Nếu đã là list hoặc kiểu dữ liệu khác, trả về nguyên bản
    return data

# BƯỚC MỚI VÀ QUAN TRỌNG: Áp dụng hàm chuyển đổi an toàn
df['parent_labels'] = df['parent_labels'].apply(safe_literal_eval)

# 3. San phẳng (flatten) dữ liệu và lấy các nhãn duy nhất
all_parent_labels_flat = [label for sublist in df['parent_labels'] for label in sublist if label and label.isalnum()]
unique_parent_labels = sorted(list(set(all_parent_labels_flat)))

print("\nĐã lấy xong các nhãn 'parent_labels' duy nhất.")
print(f"Danh sách các nhãn duy nhất: {unique_parent_labels}")

# 4. Tạo embeddings cho các nhãn và chuẩn hóa
print("\nĐang tạo embeddings cho các 'parent_labels' và chuẩn hóa...")
parent_label_embeddings_list = model.encode(
    unique_parent_labels,
    show_progress_bar=True,
    normalize_embeddings=True # Thêm tham số này để chuẩn hóa vector
)
print("Đã tạo xong và chuẩn hóa embeddings.")

# 5. Tạo dictionary ánh xạ nhãn với embeddings đã chuẩn hóa
parent_label_dict = {name: embedding for name, embedding in zip(unique_parent_labels, parent_label_embeddings_list)}
print("\nĐã tạo xong dictionary cho parent_labels.")
print(f"Số lượng entries trong dictionary: {len(parent_label_dict)}")


Đã lấy xong các nhãn 'parent_labels' duy nhất.
Danh sách các nhãn duy nhất: ['cs', 'econ', 'eess', 'math', 'nlin', 'physics', 'stat']

Đang tạo embeddings cho các 'parent_labels' và chuẩn hóa...


NameError: name 'model' is not defined

In [ ]:
type(parent_label_dict)

dict

In [ ]:
# In ra tổng số items
so_luong_item = len(parent_label_dict)
print(f"Tổng số item trong dict: {so_luong_item}")

Tổng số item trong dict: 7


In [ ]:
counter = 0
for key, value in parent_label_dict.items():
    if counter < 20:
        print(f"Key: {key} -> Value: {value.shape}")
        counter += 1
    else:
        break # Dừng vòng lặp sau khi đã in đủ 5 item

Key: cs -> Value: (384,)
Key: econ -> Value: (384,)
Key: eess -> Value: (384,)
Key: math -> Value: (384,)
Key: nlin -> Value: (384,)
Key: physics -> Value: (384,)
Key: stat -> Value: (384,)


In [ ]:
parent_label_dict.items()

dict_items([('cs', array([-6.09161854e-02,  4.02116217e-02, -9.09962207e-02,  7.09643587e-02,
       -4.79100039e-03, -6.24113493e-02,  9.76989120e-02,  2.21683607e-02,
        1.96847115e-02,  4.15325351e-03, -4.99133207e-02, -8.47239122e-02,
        4.57276665e-02, -4.64608409e-02, -2.60798857e-02, -2.12841271e-03,
        2.10789796e-02, -4.10184637e-02, -6.56680167e-02, -4.45803292e-02,
       -1.33160159e-01, -8.48432407e-02, -4.96465974e-02,  6.83533261e-03,
       -2.70084161e-02,  4.89896089e-02, -1.69768091e-02,  3.46207470e-02,
        2.91205361e-03, -1.05241172e-01, -6.99328855e-02,  9.88640357e-03,
        1.06534071e-01,  2.47874148e-02, -5.16984612e-02, -3.65980677e-02,
        1.19524542e-02, -2.19083428e-02, -9.03326422e-02,  1.04752863e-02,
       -4.12750989e-02, -8.27054027e-03,  2.44851969e-02,  2.30973568e-02,
        4.61381637e-02, -2.73780245e-03, -1.56594254e-02, -1.44659886e-02,
       -2.54759695e-02,  8.93610120e-02, -6.37395382e-02, -4.27874550e-02,
      

### lưu lại ket qua

# 1.Tải .pkl đã xử lý từ Google Drive

Phần code này thực hiện việc tải các dữ liệu đã được xử lý và lưu trước đó dưới dạng file `.pkl` trên Google Drive. Điều này giúp bạn tiếp tục công việc mà không cần chạy lại các bước tiền xử lý và tạo embeddings từ đầu.

1.  **Mount Google Drive:**
    *   Kiểm tra và mount (gắn kết) Google Drive vào môi trường Colab để có thể truy cập các file đã lưu. Nếu Drive đã được mount, thông báo sẽ hiển thị.

2.  **Chỉ định đường dẫn file:**
    *   Biến `file_path` chứa đường dẫn cụ thể đến file `30Kdata.pkl` trên Google Drive của bạn. Đảm bảo đường dẫn này là chính xác.

3.  **Tải dữ liệu từ file `.pkl`:**
    *   Sử dụng thư viện `pickle` để mở và đọc file nhị phân (`'rb'`).
    *   Hàm `pickle.load(f)` sẽ deserialze (giải mã) dữ liệu từ file và tải nó vào dictionary `loaded_data`.

4.  **Gán dữ liệu vào các biến:**
    *   Các trường ('abstract_embeddings', 'parent_label_embeddings_dict', 'original_abstracts', 'original_parent_labels', 'original_children_labels') từ dictionary `loaded_data` được gán vào các biến riêng biệt để dễ dàng sử dụng trong các bước tiếp theo.

5.  **Kiểm tra và báo cáo:**
    *   In ra thông báo xác nhận việc tải file thành công và liệt kê các biến đã được tải.
    *   Xử lý trường hợp không tìm thấy file (`FileNotFoundError`) hoặc các lỗi khác trong quá trình tải.

### Tái tạo Dictionary `processed_data`

Sau khi tải các biến dữ liệu riêng lẻ, đoạn code này tập hợp lại chúng vào một dictionary duy nhất có tên `processed_data`. Cấu trúc này giống với dictionary đã được lưu ban đầu, giúp giữ tính nhất quán và dễ dàng truy cập dữ liệu theo cùng một cách.

1.  **Tạo Dictionary mới:**
    *   Một dictionary mới được tạo ra với các key tương ứng ('abstract_embeddings', 'parent_label_embeddings_dict', 'original_abstracts', 'original_parent_labels', 'original_children_labels').
    *   Các biến đã tải ở bước trước được gán làm value cho các key này.

2.  **Kiểm tra Dictionary đã tái tạo:**
    *   In ra thông báo xác nhận dictionary `processed_data` đã được tạo lại thành công và liệt kê các key có trong đó.

Các bước này đảm bảo rằng tất cả dữ liệu cần thiết đã sẵn sàng trong môi trường làm việc để tiếp tục các phân tích hoặc xây dựng mô hình, ví dụ như sử dụng cho Nearest Neighbor Search (ANN) hoặc các thuật toán học máy khác.

In [1]:
import pickle
from google.colab import drive

# Mount Google Drive if not already mounted
try:
    drive.mount('/content/drive')
except:
    print("Drive already mounted.")

# Define the path to the pickle file
file_path = '/content/drive/MyDrive/AIO/30Kdata.pkl'

# Load the data from the pickle file
try:
    with open(file_path, 'rb') as f:
        loaded_data = pickle.load(f)

    # Assign the loaded data to the respective variables
    abstract_embeddings = loaded_data['abstract_embeddings']
    parent_label_embeddings_dict = loaded_data['parent_label_embeddings_dict']
    original_abstracts = loaded_data['original_abstracts']
    original_parent_labels = loaded_data['original_parent_labels']
    original_children_labels = loaded_data['original_children_labels']

    print(f"Successfully loaded data from: {file_path}")
    print(f"Loaded variables: {list(loaded_data.keys())}")

except FileNotFoundError:
    print(f"Error: File not found at {file_path}. Please check the path.")
except Exception as e:
    print(f"Error loading data: {e}")

Mounted at /content/drive
Successfully loaded data from: /content/drive/MyDrive/AIO/30Kdata.pkl
Loaded variables: ['abstract_embeddings', 'parent_label_embeddings_dict', 'original_abstracts', 'original_parent_labels', 'original_children_labels']


In [2]:
# Recreate the processed_data dictionary from the loaded variables
processed_data = {
    'abstract_embeddings': abstract_embeddings,
    'parent_label_embeddings_dict': parent_label_embeddings_dict,
    'original_abstracts': original_abstracts,
    'original_parent_labels': original_parent_labels,
    'original_children_labels': original_children_labels
}

print("\nĐã tái tạo dictionary processed_data.")
print(f"Các trường trong processed_data: {list(processed_data.keys())}")


Đã tái tạo dictionary processed_data.
Các trường trong processed_data: ['abstract_embeddings', 'parent_label_embeddings_dict', 'original_abstracts', 'original_parent_labels', 'original_children_labels']


## 1.1 Kiểm tra file.pkl

In [3]:
type(processed_data)

dict

In [4]:
# In dữ liệu của 25 mẫu đầu tiên từ processed_data
num_samples_to_show = 25

print(f"--- Hiển thị dữ liệu của {num_samples_to_show} mẫu đầu tiên từ 'processed_data' ---")

# Kiểm tra xem processed_data có các key cần thiết không
if all(key in processed_data for key in ['original_abstracts', 'original_parent_labels', 'original_children_labels']):
    for i in range(num_samples_to_show):
        if i < len(processed_data['original_abstracts']): # Đảm bảo không vượt quá số lượng mẫu thực tế
            print(f"\n--- Mẫu {i+1} ---")
            # Hiển thị một phần abstract để ngắn gọn
            print(f"Abstract: {processed_data['original_abstracts'][i][:200]}...")
            print(f"Nhãn Cha: {processed_data['original_parent_labels'][i]}")
            print(f"Nhãn Con: {processed_data['original_children_labels'][i]}")
            print("-" * 30)
        else:
            print(f"\nĐã hết mẫu sau {i} mẫu.")
            break
else:
    print("\nLỗi: Dictionary 'processed_data' không chứa các trường cần thiết (original_abstracts, original_parent_labels, original_children_labels).")
    print(f"Các trường hiện có: {list(processed_data.keys())}")

--- Hiển thị dữ liệu của 25 mẫu đầu tiên từ 'processed_data' ---

--- Mẫu 1 ---
Abstract:   By studying a tachyon field on the DGP brane model, in order to embed the 4D
standard Friedmann equation with a brane tachyon field in 5D bulk, the metric
of the 5D spacetime is presented. Then, ado...
Nhãn Cha: ['hep-th']
Nhãn Con: hep-th
------------------------------

--- Mẫu 2 ---
Abstract:   Gravitational waves in the sensitivity band of ground-based detectors can be
emitted by a number of astrophysical sources, including not only binary
coalescences, but also individual spinning neutro...
Nhãn Cha: ['gr-qc', 'astro-ph']
Nhãn Con: gr-qc astro-ph.IM
------------------------------

--- Mẫu 3 ---
Abstract:   With the Sturm-Liouville analytical and numerical methods, we investigate the
behaviors of the holographic superconductors by introducing a complex charged
scalar field coupled with a Power-Maxwell ...
Nhãn Cha: ['hep-th', 'gr-qc']
Nhãn Con: hep-th gr-qc
------------------------------

---

In [5]:
# Khám phá cấu trúc của dictionary processed_data
print("--- Khám phá dictionary 'processed_data' ---")
print(f"Số lượng cặp key-value trong dictionary: {len(processed_data)}")

print("\n--- Thông tin chi tiết về từng trường trong dictionary ---")
for key, value in processed_data.items():
    print(f"Key: {key}")
    # Kiểm tra xem có phải là NumPy array hoặc đối tượng tương tự không
    if hasattr(value, 'shape'):
        print(f"  - Kiểu dữ liệu: {type(value)}")
        print(f"  - Kích thước (shape): {value.shape}")
    # Nếu là list hoặc dict, in ra độ dài của chúng
    elif isinstance(value, (list, dict)):
        print(f"  - Kiểu dữ liệu: {type(value)}")
        print(f"  - Số lượng phần tử (len): {len(value)}")
    # Cho các trường hợp khác
    else:
        print(f"  - Kiểu dữ liệu: {type(value)}")
        # Tránh in giá trị lớn, chỉ in đại diện nếu cần
        # print(f"  - Giá trị (ví dụ): {str(value)[:100]}...")

    print("-" * 20) # Thêm dòng phân cách

--- Khám phá dictionary 'processed_data' ---
Số lượng cặp key-value trong dictionary: 5

--- Thông tin chi tiết về từng trường trong dictionary ---
Key: abstract_embeddings
  - Kiểu dữ liệu: <class 'numpy.ndarray'>
  - Kích thước (shape): (60000, 384)
--------------------
Key: parent_label_embeddings_dict
  - Kiểu dữ liệu: <class 'dict'>
  - Số lượng phần tử (len): 7
--------------------
Key: original_abstracts
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------
Key: original_parent_labels
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------
Key: original_children_labels
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------


# 1. Tiền xử lý dữ liệu ban đầu về dạng json cha con

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import pandas as pd

# Dán đường dẫn của file vào đây
file_path = '/content/drive/MyDrive/AIO/top 30 main categories.csv'
# Thay '/content/drive/MyDrive/...' bằng đường dẫn thật của bạn

df = pd.read_csv(file_path)

# In 5 dòng đầu tiên của file để kiểm tra
df.head(25)

authors  \
0        Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1   David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2              Jiliang Jing, Qiyuan Pan, Songbai Chen   
3   R. Hole\v{n}\'ak, S. Lohmann, F. Sekula and D....   
4                 Szabolcs Horv\'at and Carl D. Modes   
5                          A. Hopper and P. F. Barker   
6                   Zeyang Liao and M. Suhail Zubairy   
7   Christian Majenz, Maris Ozols, Christian Schaf...   
8   Sergey Khaykin, Bernard Legras, Silvia Bucci, ...   
9                    Wenchao Ge and M. Suhail Zubairy   
10              Grzegorz Wilk and Zbigniew Wlodarczyk   
11                 Alexander Grant, Eanna E. Flanagan   
12                                      M. A. Sol\'is   
13                                   Kaushik Majumdar   
14  Visakan Balakumar, Rafael P. Bernar, Luis C. B...   
15  G.G.Adamian, N.V.Antonenko, A.Diaz-Torres, W.S...   
16  BESIII Collaboration: M. Ablikim, M. N. Achaso...   
17  Sung Mook Lee, Dhong Yeon Cheong, Sang Chul Hy...   
18                                         S. Mignemi   
19            Sanjib Kumar Agarwalla, Pilar Hernandez   
20                          T-Y. Saito and I.R. Afnan   
21  S.J. Dong, T. Draper, I. Horvath, Nilmani Math...   
22  A.K. Karlis, P. K. Papachristou, F. K. Diakono...   
23  Lucas Tendick, Hermann Kampermann, Dagmar Bru{...   
24                                ATLAS Collaboration   

                                                title  \
0   Power-law cosmological solution derived from D...   
1   PyFstat: a Python package for continuous gravi...   
2   Holographic Superconductors with Power-Maxwell...   
3   Simultaneous assessment of energy, charge stat...   
4   Connectedness matters: Construction and exact ...   
5   A hybrid quantum system formed by trapping ato...   
6   Quantum state preparation by a shaped photon p...   
7             Local simultaneous state discrimination   
8   The 2019/20 Australian wildfires generated a p...   
9   Dark-State Optical Potential Barriers with Nan...   
10  Self-similarity in jet events following from p...   
11  Non-conservation of Carter in black hole space...   
12  Tamm's surface states and Bose-Einstein conden...   
13  Fourier Uniformity: An Useful Tool for Analyzi...   
14  Quantum superradiance on static black hole spa...   
15  Dynamical restriction for a growing neck due t...   
16  Measurement of the Cross Section for $e^{+}e^{...   
17  Festina-Lente Bound on Higgs Vacuum Structure ...   
18     Doubly Special Relativity and Finsler geometry   
19  Probing the Neutrino Mass Hierarchy with Super...   
20          How important is the three-nucleon force?   
21  Empirical Baye's Method and Test in Very Light...   
22  Hyperacceleration in a stochastic Fermi-Ulam m...   
23  Quantifying necessary quantum resources for no...   
24  Search for direct top squark pair production i...   

                                             abstract  \
0     By studying a tachyon field on the DGP brane...   
1     Gravitational waves in the sensitivity band ...   
2     With the Sturm-Liouville analytical and nume...   
3     We demonstrate simultaneous measurements of ...   
4     We describe a new method for the random samp...   
5     Near-field, radially symmetric optical poten...   
6     We propose a method to deterministically pre...   
7     Quantum state discrimination is one of the m...   
8     The Australian bushfires around the turn of ...   
9     Optical potentials have been a versatile too...   
10    Using a Tsallis nonextensive approach, we si...   
11    Freely falling point particles in the vicini...   
12    We calculate and discuss the effects on the ...   
13    If two signals are phase synchronous then th...   
14    We study the quantum analogue of the classic...   
15    Mass parameters for collective variables of ...   
16    Based on electron-positron collision data co...   
17    The recently suggested Festina-Lente (FL) bo... 

--- Khám phá dictionary 'processed_data' ---
Số lượng cặp key-value trong dictionary: 5

--- Thông tin chi tiết về từng trường trong dictionary ---
Key: abstract_embeddings
  - Kiểu dữ liệu: <class 'numpy.ndarray'>
  - Kích thước (shape): (60000, 384)
--------------------
Key: parent_label_embeddings_dict
  - Kiểu dữ liệu: <class 'dict'>
  - Số lượng phần tử (len): 7
--------------------
Key: original_abstracts
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------
Key: original_parent_labels
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------
Key: original_children_labels
  - Kiểu dữ liệu: <class 'list'>
  - Số lượng phần tử (len): 60000
--------------------


```python
import pandas as pd
import ast
import json

# ===== Cấu hình =====
data_path = "arxiv_perfectly_balanced.csv"  # đổi nếu cần
out_path  = "dropped_col_df.csv"
cols = ["authors", "title", "abstract", "categories", "parent_labels", "child_labels"]

# ===== Đọc file chỉ lấy cột cần =====
df = pd.read_csv(data_path, usecols=cols)

# ===== Chuẩn hoá 2 cột list (nếu đang lưu dạng chuỗi list trong CSV) =====
for col in ["parent_labels", "child_labels"]:
    df[col] = df[col].apply(
        lambda x: x if isinstance(x, list)
        else ([] if pd.isna(x) else ast.literal_eval(str(x)))
    )

# ===== Tạo cột json_categories: với mỗi cha, quét toàn bộ con để lấy con có tiền tố đúng =====
df["json_categories"] = df.apply(
    lambda r: {
        p: [c for c in (r["child_labels"] or [])
             if isinstance(c, str) and (c == p or c.startswith(p + ".") or c.startswith(p + "-"))]
        for p in (r["parent_labels"] or [])
    },
    axis=1
)

# Tuỳ chọn: nếu muốn lưu ra CSV ở dạng chuỗi JSON cho cột json_categories
df["json_categories"] = df["json_categories"].apply(lambda d: json.dumps(d, ensure_ascii=False))

# ===== Lưu kết quả =====
df_processed=df.drop(columns=["categories", "parent_labels", "child_labels"])
df_processed.to_csv(out_path, index=False)  # bỏ tham số file_name khỏi đoạn lệnh to_csv
```

In [95]:
import pandas as pd
import ast
import json

# ===== Cấu hình =====
data_path = "/content/drive/MyDrive/AIO/top 30 main categories.csv"  # Đường dẫn file trên Google Drive
out_path  = "processed_arxiv_data.csv" # Tên file output (sẽ lưu trong thư mục hiện tại của Colab)
# Điều chỉnh cột theo yêu cầu: categories là con, main_categories là cha
cols = ["authors", "title", "abstract", "categories", "main_categories"]

# ===== Đọc file chỉ lấy cột cần =====
df = pd.read_csv(data_path, usecols=cols)

# # Đổi tên cột theo yêu cầu - Bỏ qua bước đổi tên cột 'categories'
df = df.rename(columns={
    'main_categories': 'parent_labels'
})

# ===== Chuẩn hoá 2 cột list (nếu đang lưu dạng chuỗi list trong CSV) =====
# Cần kiểm tra kiểu dữ liệu trước khi áp dụng ast.literal_eval
def safe_literal_eval(data, col_name):
    if isinstance(data, str):
        try:
            return ast.literal_eval(data)
        except (ValueError, SyntaxError):
            # Nếu không phải chuỗi list hợp lệ, thử tách bằng dấu cách cho nhãn con ('categories')
            # Chú ý: Đây có thể cần điều chỉnh tùy vào format dữ liệu gốc không phải list string
            return data.split() if col_name == 'categories' and isinstance(data, str) else []
    elif isinstance(data, list):
        return data # Nếu đã là list, giữ nguyên
    else:
        return [] # Xử lý các kiểu dữ liệu khác (ví dụ: NaN)

# Áp dụng hàm chuẩn hóa cho cả hai cột, truyền tên cột để xử lý 'categories' khác
df['parent_labels'] = df['parent_labels'].apply(lambda x: safe_literal_eval(x, 'parent_labels'))
df['categories'] = df['categories'].apply(lambda x: safe_literal_eval(x, 'categories'))


# ===== Tạo cột json_categories: với mỗi cha, quét toàn bộ con để lấy con có tiền tố đúng =====
# Sử dụng tên cột 'categories' (nhãn con)
df["json_categories"] = df.apply(
    lambda r: {
        p: [c for c in (r["categories"] or [])
             if isinstance(c, str) and (c == p or c.startswith(p + ".") or c.startswith(p + "-"))]
        for p in (r["parent_labels"] or [])
    },
    axis=1
)

# Tuỳ chọn: nếu muốn lưu ra CSV ở dạng chuỗi JSON cho cột json_categories
df["json_categories"] = df["json_categories"].apply(lambda d: json.dumps(d, ensure_ascii=False))

# ===== Lưu kết quả =====
# Giữ lại các cột cần thiết sau khi xử lý, sử dụng 'categories' cho nhãn con
df_processed = df[['authors', 'title', 'abstract', 'categories', 'json_categories']]
df_processed.to_csv(out_path, index=False)

print(f"Đã xử lý và lưu file '{out_path}' thành công.")
print(df_processed.head()) # In vài dòng đầu của df_processed để kiểm tra

Đã xử lý và lưu file 'processed_arxiv_data.csv' thành công.
                                             authors  \
0       Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1  David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2             Jiliang Jing, Qiyuan Pan, Songbai Chen   
3  R. Hole\v{n}\'ak, S. Lohmann, F. Sekula and D....   
4                Szabolcs Horv\'at and Carl D. Modes   

                                               title  \
0  Power-law cosmological solution derived from D...   
1  PyFstat: a Python package for continuous gravi...   
2  Holographic Superconductors with Power-Maxwell...   
3  Simultaneous assessment of energy, charge stat...   
4  Connectedness matters: Construction and exact ...   

                                            abstract  \
0    By studying a tachyon field on the DGP brane...   
1    Gravitational waves in the sensitivity band ...   
2    With the Sturm-Liouville analytical and nume...   
3    We demonstrate simultaneous measureme

In [9]:
df_processed.head(100)

authors  \
0        Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1   David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2              Jiliang Jing, Qiyuan Pan, Songbai Chen   
3   R. Hole\v{n}\'ak, S. Lohmann, F. Sekula and D....   
4                 Szabolcs Horv\'at and Carl D. Modes   
..                                                ...   
95                    Chian-Shu Chen and Chia-Min Lin   
96                       Fady Bishara and Zhuoni Qian   
97  M. Bruni, S. Matarrese, S. Mollerach, and S. S...   
98                                         Ilia Musco   
99  Henning Bahl, Thomas Biek\"otter, Sven Heineme...   

                                                title  \
0   Power-law cosmological solution derived from D...   
1   PyFstat: a Python package for continuous gravi...   
2   Holographic Superconductors with Power-Maxwell...   
3   Simultaneous assessment of energy, charge stat...   
4   Connectedness matters: Construction and exact ...   
..                                                ...   
95  Cosmon as the Modulon: Non-Gaussianity from Da...   
96  Complementary constraints on $Zb\bar{b}$ coupl...   
97  On Relativistic Perturbations of Second and Hi...   
98  Threshold for primordial black holes: Dependen...   
99  HiggsTools: BSM scalar phenomenology with new ...   

                                             abstract  \
0     By studying a tachyon field on the DGP brane...   
1     Gravitational waves in the sensitivity band ...   
2     With the Sturm-Liouville analytical and nume...   
3     We demonstrate simultaneous measurements of ...   
4     We describe a new method for the random samp...   
..                                                ...   
95    In this paper, we show that the idea of grow...   
96    We propose a new strategy to probe the $Z$ b...   
97    We present the results of a study of the gau...   
98    Primordial black holes may have formed in th...   
99    The codes HiggsBounds and HiggsSignals compa...   

                          categories  \
0                           [hep-th]   
1               [gr-qc, astro-ph.IM]   
2                    [hep-th, gr-qc]   
3                  [physics.ins-det]   
4   [physics.soc-ph, cs.DM, math.CO]   
..                               ...   
95     [hep-ph, astro-ph.CO, hep-th]   
96                  [hep-ph, hep-ex]   
97                 [gr-qc, astro-ph]   
98              [gr-qc, astro-ph.CO]   
99                          [hep-ph]   

                                      json_categories  
0                              {"hep-th": ["hep-th"]}  
1   {"gr-qc": ["gr-qc"], "astro-ph": ["astro-ph.IM"]}  
2          {"hep-th": ["hep-th"], "gr-qc": ["gr-qc"]}  
3                    {"physics": ["physics.ins-det"]}  
4                     {"physics": ["physics.soc-ph"]}  
..                                                ...  
95  {"hep-ph": ["hep-ph"], "astro-ph": ["astro-ph....  
96       {"hep-ph": ["hep-ph"], "hep-ex": ["hep-ex"]}  
97     {"gr-qc": ["gr-qc"], "astro-ph": ["astro-ph"]}  
98  {"gr-qc": ["gr-qc"], "astro-ph": ["astro-ph.CO"]}  
99                             {"hep-ph": ["hep-ph"]}  

[100 rows x 5 columns]

In [96]:
# Chọn chỉ cột 'json_categories'
json_categories_df = df_processed[['json_categories']]

# Định nghĩa tên file output
output_json_path = "/content/drive/MyDrive/AIO/json_categories_only.csv"

# Lưu DataFrame mới vào file CSV
json_categories_df.to_csv(output_json_path, index=False)

print(f"Đã lưu file '{output_json_path}' chỉ với cột 'json_categories'.")

Đã lưu file '/content/drive/MyDrive/AIO/json_categories_only.csv' chỉ với cột 'json_categories'.


In [11]:
len(df_processed)

60000

# Thống kê nhãn cha, nhãn con từ json_categories_df

In [11]:
df_processed = pd.read_csv('/content/drive/MyDrive/AIO/json_categories_only.csv')

In [97]:
import json
import pandas as pd
from collections import Counter, defaultdict

def analyze_arxiv_categories_extended(json_categories_data):
    """
    Phân tích json_categories:
    Trả về:
      - parent_count: Counter({parent: số_entry_chứa_parent})
      - parent_unique_subcats: dict({parent: set(nhãn_con_unique)})
      - parent_child_counts: dict({parent: Counter({child: tần_suất})})
      - total_entries: tổng số entry hợp lệ (parse JSON OK)
    """
    parent_count = Counter()
    parent_unique_subcats = defaultdict(set)
    parent_child_counts = defaultdict(Counter)
    total_entries = 0

    for line in json_categories_data:
        if not isinstance(line, (str, dict)) or (isinstance(line, str) and not line.strip()):
            continue

        try:
            categories = json.loads(line.strip()) if isinstance(line, str) else line
            if not isinstance(categories, dict):
                continue
            total_entries += 1

            for parent, subcats in categories.items():
                parent_count[parent] += 1
                # Chuẩn hoá subcats về list
                if subcats is None:
                    continue
                if not isinstance(subcats, list):
                    # Nếu là string "['a','b']" thì cố parse
                    try:
                        parsed = json.loads(subcats)
                        subcats = parsed if isinstance(parsed, list) else []
                    except Exception:
                        subcats = []

                for child in subcats:
                    parent_unique_subcats[parent].add(child)
                    parent_child_counts[parent][child] += 1

        except json.JSONDecodeError:
            # Bỏ qua JSON hỏng
            continue

    return parent_count, parent_unique_subcats, parent_child_counts, total_entries


def display_results_extended(parent_count, parent_unique_subcats, parent_child_counts, total_entries, top_k_children=10):
    """
    Hiển thị:
      - Tổng quan
      - Bảng nhãn cha: số lần xuất hiện + số nhãn con unique
      - Top-k nhãn con theo từng nhãn cha
    """
    total_parent_occurrences = sum(parent_count.values())
    print("📊 TỔNG QUAN")
    print("="*50)
    print(f"Tổng số entries hợp lệ: {total_entries}")
    print(f"Số nhãn cha duy nhất: {len(parent_count)}")
    print(f"Tổng lượt xuất hiện nhãn cha: {total_parent_occurrences}\n")

    # Bảng nhãn cha
    rows = []
    for p, c in parent_count.items():
        rows.append({
            "nhãn_cha": p,
            "số_lần_xuất_hiện": c,
            "số_nhãn_con_unique": len(parent_unique_subcats.get(p, set()))
        })
    df_parents = pd.DataFrame(rows).sort_values("số_lần_xuất_hiện", ascending=False)
    print("📋 THỐNG KÊ NHÃN CHA")
    print(df_parents.head(20).to_string(index=False))
    print()

    # Top-k nhãn con theo từng nhãn cha
    print("🔎 TOP nhãn con theo từng nhãn cha")
    print("="*50)
    for p in df_parents["nhãn_cha"].tolist():
        cnt = parent_child_counts.get(p, Counter())
        if not cnt:
            continue
        total_sub = sum(cnt.values())
        top_items = cnt.most_common(top_k_children)
        top_list = [f"{ch} ({n}, {n/total_sub*100:.1f}%)" for ch, n in top_items]
        print(f"- {p}: {'; '.join(top_list)}")
    print()

    # Thống kê tóm tắt
    print("📈 TÓM TẮT")
    print("="*50)
    for i, row in enumerate(df_parents.head(5).itertuples(index=False), 1):
        pct = row.số_lần_xuất_hiện / total_parent_occurrences * 100 if total_parent_occurrences else 0
        print(f"  {i}. {row.nhãn_cha}: {row.số_lần_xuất_hiện} lần ({pct:.1f}%)")


def results_to_dataframes(parent_count, parent_unique_subcats, parent_child_counts):
    """
    Chuyển kết quả sang 2 DataFrame:
      - df_parents: 1 dòng/nhãn_cha
      - df_parent_child: 1 dòng/(nhãn_cha, nhãn_con) với tần suất
    """
    df_parents = pd.DataFrame([
        {
            "nhãn_cha": p,
            "số_lần_xuất_hiện": parent_count[p],
            "số_nhãn_con_unique": len(parent_unique_subcats.get(p, set())),
            "danh_sách_nhãn_con_unique": ", ".join(sorted(parent_unique_subcats.get(p, set())))
        }
        for p in parent_count.keys()
    ]).sort_values("số_lần_xuất_hiện", ascending=False)

    rows = []
    for p, counter in parent_child_counts.items():
        for ch, n in counter.items():
            rows.append({"nhãn_cha": p, "nhãn_con": ch, "tần_suất": n})
    df_parent_child = pd.DataFrame(rows).sort_values(["nhãn_cha", "tần_suất"], ascending=[True, False])

    return df_parents, df_parent_child


def load_json_categories_column(file_path, col="json_categories"):
    """Đọc CSV và trả về list cột json_categories (bỏ NaN)."""
    df = pd.read_csv(file_path)
    if col not in df.columns:
        raise KeyError(f"Không tìm thấy cột '{col}' trong file.")
    return df[col].dropna().tolist()


# ============================
# Ví dụ chạy (điền lại file_path của bạn)
# ============================
if __name__ == "__main__":
    print("🚀 BẮT ĐẦU PHÂN TÍCH DỮ LIỆU ARXIV")
    file_path = "/content/drive/MyDrive/AIO/json_categories_only.csv"  # đổi theo đường dẫn của bạn
    data = load_json_categories_column(file_path)

    print("🔄 Đang phân tích dữ liệu...")
    parent_count, parent_unique_subcats, parent_child_counts, total_entries = analyze_arxiv_categories_extended(data)

    print("✅ Hoàn thành phân tích!\n")
    display_results_extended(parent_count, parent_unique_subcats, parent_child_counts, total_entries, top_k_children=None)

    # Xuất CSV nếu cần
    df_parents, df_parent_child = results_to_dataframes(parent_count, parent_unique_subcats, parent_child_counts)
    df_parents.to_csv("arxiv_parent_summary.csv", index=False, encoding="utf-8")
    df_parent_child.to_csv("arxiv_parent_child_counts.csv", index=False, encoding="utf-8")
    print("💾 Đã lưu: arxiv_parent_summary.csv & arxiv_parent_child_counts.csv")


🚀 BẮT ĐẦU PHÂN TÍCH DỮ LIỆU ARXIV
🔄 Đang phân tích dữ liệu...
✅ Hoàn thành phân tích!

📊 TỔNG QUAN
Tổng số entries hợp lệ: 60000
Số nhãn cha duy nhất: 20
Tổng lượt xuất hiện nhãn cha: 82000

📋 THỐNG KÊ NHÃN CHA
nhãn_cha  số_lần_xuất_hiện  số_nhãn_con_unique
  hep-th             10000                   1
  hep-ph             10000                   1
   gr-qc              8000                   1
astro-ph              6000                   7
cond-mat              6000                  10
quant-ph              6000                   1
 physics              4000                  22
 nucl-th              4000                   1
  hep-ex              4000                   1
    math              4000                  33
 hep-lat              2000                   1
   q-bio              2000                  10
    nlin              2000                   5
    econ              2000                   3
      cs              2000                  39
    eess              2000           

```
🚀 BẮT ĐẦU PHÂN TÍCH DỮ LIỆU ARXIV
🔄 Đang phân tích dữ liệu...
✅ Hoàn thành phân tích!

📊 TỔNG QUAN
==================================================
Tổng số entries hợp lệ: 60000
Số nhãn cha duy nhất: 20
Tổng lượt xuất hiện nhãn cha: 82000

📋 THỐNG KÊ NHÃN CHA
nhãn_cha  số_lần_xuất_hiện  số_nhãn_con_unique
  hep-th             10000                   1
  hep-ph             10000                   1
   gr-qc              8000                   1
astro-ph              6000                   7
cond-mat              6000                  10
quant-ph              6000                   1
 physics              4000                  22
 nucl-th              4000                   1
  hep-ex              4000                   1
    math              4000                  33
 hep-lat              2000                   1
   q-bio              2000                  10
    nlin              2000                   5
    econ              2000                   3
      cs              2000                  39
    eess              2000                   4
 math-ph              2000                   1
 nucl-ex              2000                   1
    stat              2000                   6
   q-fin              2000                   9

🔎 TOP nhãn con theo từng nhãn cha
==================================================
- hep-th: hep-th (10000, 100.0%)
- hep-ph: hep-ph (10000, 100.0%)
- gr-qc: gr-qc (8000, 100.0%)
- astro-ph: astro-ph.CO (2416, 35.9%); astro-ph.HE (1462, 21.7%); astro-ph (1330, 19.8%); astro-ph.GA (519, 7.7%); astro-ph.SR (512, 7.6%); astro-ph.IM (282, 4.2%); astro-ph.EP (208, 3.1%)
- cond-mat: cond-mat.stat-mech (1535, 21.2%); cond-mat.str-el (1523, 21.0%); cond-mat.mes-hall (1381, 19.1%); cond-mat.mtrl-sci (641, 8.9%); cond-mat.quant-gas (486, 6.7%); cond-mat.supr-con (486, 6.7%); cond-mat.other (384, 5.3%); cond-mat.dis-nn (290, 4.0%); cond-mat.soft (261, 3.6%); cond-mat (250, 3.5%)
- quant-ph: quant-ph (6000, 100.0%)
- physics: physics.optics (1351, 27.9%); physics.atom-ph (716, 14.8%); physics.chem-ph (357, 7.4%); physics.ins-det (279, 5.8%); physics.comp-ph (260, 5.4%); physics.flu-dyn (234, 4.8%); physics.app-ph (218, 4.5%); physics.hist-ph (215, 4.4%); physics.plasm-ph (167, 3.4%); physics.soc-ph (157, 3.2%); physics.bio-ph (132, 2.7%); physics.gen-ph (121, 2.5%); physics.class-ph (112, 2.3%); physics.data-an (98, 2.0%); physics.acc-ph (74, 1.5%); physics.med-ph (62, 1.3%); physics.ao-ph (56, 1.2%); physics.geo-ph (55, 1.1%); physics.ed-ph (54, 1.1%); physics.space-ph (47, 1.0%); physics.atm-clus (43, 0.9%); physics.pop-ph (33, 0.7%)
- nucl-th: nucl-th (4000, 100.0%)
- hep-ex: hep-ex (4000, 100.0%)
- math: math-ph (2070, 26.8%); math.MP (2070, 26.8%); math.AP (376, 4.9%); math.PR (296, 3.8%); math.CO (275, 3.6%); math.AG (229, 3.0%); math.DG (215, 2.8%); math.DS (188, 2.4%); math.NT (170, 2.2%); math.FA (156, 2.0%); math.OC (153, 2.0%); math.RT (143, 1.9%); math.NA (132, 1.7%); math.SP (121, 1.6%); math.CA (109, 1.4%); math.RA (94, 1.2%); math.GR (94, 1.2%); math.QA (93, 1.2%); math.GT (86, 1.1%); math.OA (84, 1.1%); math.CV (83, 1.1%); math.SG (67, 0.9%); math.ST (64, 0.8%); math.LO (63, 0.8%); math.AT (61, 0.8%); math.AC (55, 0.7%); math.MG (53, 0.7%); math.CT (33, 0.4%); math.GM (28, 0.4%); math.KT (24, 0.3%); math.GN (15, 0.2%); math.HO (12, 0.2%); math.IT (11, 0.1%)
- hep-lat: hep-lat (2000, 100.0%)
- q-bio: q-bio.PE (570, 23.6%); q-bio.QM (445, 18.4%); q-bio.NC (422, 17.5%); q-bio.BM (241, 10.0%); q-bio.MN (199, 8.2%); q-bio.GN (167, 6.9%); q-bio.TO (109, 4.5%); q-bio.CB (106, 4.4%); q-bio.SC (81, 3.4%); q-bio.OT (75, 3.1%)
- nlin: nlin.CD (760, 34.1%); nlin.SI (576, 25.8%); nlin.PS (510, 22.8%); nlin.AO (317, 14.2%); nlin.CG (69, 3.1%)
- econ: econ.GN (862, 41.6%); econ.EM (719, 34.7%); econ.TH (492, 23.7%)
- cs: cs.LG (551, 18.6%); cs.CV (400, 13.5%); cs.AI (302, 10.2%); cs.CL (222, 7.5%); cs.IT (173, 5.8%); cs.CR (125, 4.2%); cs.RO (97, 3.3%); cs.DS (91, 3.1%); cs.NI (78, 2.6%); cs.DC (77, 2.6%); cs.IR (67, 2.3%); cs.LO (64, 2.2%); cs.CY (59, 2.0%); cs.SI (59, 2.0%); cs.SE (54, 1.8%); cs.NE (51, 1.7%); cs.HC (50, 1.7%); cs.DM (49, 1.7%); cs.SY (44, 1.5%); cs.GT (36, 1.2%); cs.SD (31, 1.0%); cs.DB (30, 1.0%); cs.CC (29, 1.0%); cs.PL (25, 0.8%); cs.CG (23, 0.8%); cs.DL (22, 0.7%); cs.FL (20, 0.7%); cs.CE (20, 0.7%); cs.AR (17, 0.6%); cs.MA (16, 0.5%); cs.NA (14, 0.5%); cs.SC (11, 0.4%); cs.GR (11, 0.4%); cs.PF (11, 0.4%); cs.MM (9, 0.3%); cs.ET (9, 0.3%); cs.OH (8, 0.3%); cs.MS (7, 0.2%); cs.OS (3, 0.1%)
- eess: eess.SP (733, 34.8%); eess.IV (586, 27.8%); eess.SY (522, 24.8%); eess.AS (266, 12.6%)
- math-ph: math-ph (2000, 100.0%)
- nucl-ex: nucl-ex (2000, 100.0%)
- stat: stat.ME (870, 33.2%); stat.ML (828, 31.6%); stat.AP (529, 20.2%); stat.CO (235, 9.0%); stat.TH (112, 4.3%); stat.OT (43, 1.6%)
- q-fin: q-fin.ST (477, 17.7%); q-fin.MF (359, 13.4%); q-fin.RM (321, 11.9%); q-fin.GN (317, 11.8%); q-fin.CP (314, 11.7%); q-fin.PR (305, 11.3%); q-fin.PM (264, 9.8%); q-fin.TR (218, 8.1%); q-fin.EC (114, 4.2%)

📈 TÓM TẮT
==================================================
  1. hep-th: 10000 lần (12.2%)
  2. hep-ph: 10000 lần (12.2%)
  3. gr-qc: 8000 lần (9.8%)
  4. astro-ph: 6000 lần (7.3%)
  5. cond-mat: 6000 lần (7.3%)
💾 Đã lưu: arxiv_parent_summary.csv & arxiv_parent_child_counts.csv
```

### Chỉ giữ lại các nhãn con quan trọng:
từ json_categories_df
giữ lại 50 nhãn xếp hạng cao nhất, với mỗi nhãn lấy 300 dòng data

```
giờ cần short các nhãn con từ trên xuống dưới
lấy 50 nhãn con có số lần xuất hiện cao nhất

từ data gốc, tạo df mới chỉ giữ lại 50 nhãn con cao nhất này, số lần xuất hiện của mỗi nhãn con là khoảng 300 dòng.
bằng cách;
từ nhãn cao nhất lấy 300 dòng xuất hiện nhãn con này
tiếp tục: với nhãn con tiếp theo, đếm trong data đã lấy xem số lần xuất hiện của nhãn con thứ 2 này là bao nhiêu, thiếu bao nhiêu thì lấy bù thêm các dòng data
Cứ vậy tiếp tục
```

```

Ok, mình viết sẵn đoạn code “chạy được ngay” theo đúng yêu cầu:

Lấy top 50 nhãn con phổ biến nhất từ dữ liệu gốc (df_processed['categories']).

Tạo DataFrame mới gồm tối đa 300 dòng cho mỗi nhãn con, chọn theo thứ tự từ nhãn con nhiều → ít.

Khi xét nhãn con tiếp theo, đếm số lần nhãn đó đã xuất hiện trong tập đã chọn (do bị “ăn ké” từ các nhãn trước), và chỉ lấy bù phần thiếu từ các dòng chưa được lấy.

Yêu cầu dữ liệu gốc: df_processed có cột categories là list nhãn con (nếu là chuỗi dạng list cũng được, code sẽ chuẩn hoá).
```

In [98]:
import pandas as pd
import ast
from collections import Counter, defaultdict

# --------------------------
# 1) Chuẩn hoá cột categories về List[str]
# --------------------------
def normalize_categories_col(df, col="categories"):
    if col not in df.columns:
        raise KeyError(f"Không tìm thấy cột '{col}' trong df.")
    def to_list(x):
        if isinstance(x, list):
            return x
        if pd.isna(x):
            return []
        if isinstance(x, str):
            # cố gắng parse kiểu "['a','b']" hoặc '["a","b"]'
            try:
                v = ast.literal_eval(x)
                return v if isinstance(v, list) else []
            except Exception:
                # fallback: tách theo dấu phẩy nếu cần
                return [s.strip() for s in x.split(",")] if "," in x else [x.strip()]
        return []
    out = df.copy()
    out[col] = out[col].apply(to_list)
    return out

# --------------------------
# 2) Tạo inverted index: child_label -> list chỉ số dòng
# --------------------------
def build_inverted_index(categories_series):
    inv = defaultdict(list)
    for i, cats in categories_series.items():
        for c in cats:
            inv[c].append(i)
    return inv

# --------------------------
# 3) Lấy top-K nhãn con theo tần suất xuất hiện (đếm theo số dòng có chứa nhãn đó)
# --------------------------
def top_k_child_labels(inverted_index, k=50):
    counts = {lbl: len(idxs) for lbl, idxs in inverted_index.items()}
    top = sorted(counts.items(), key=lambda x: x[1], reverse=True)[:k]
    return [lbl for lbl, _ in top], counts

# --------------------------
# 4) Chọn hàng theo chiến lược "điền quota 300/nhãn"
#     - Duy trì set chỉ số đã chọn
#     - Với mỗi nhãn: đếm đã có bao nhiêu trong set, bổ sung từ các ứng viên còn lại
# --------------------------
def select_rows_by_label_quota(df, inverted_index, labels_in_order, quota_per_label=300):
    selected = set()                         # chỉ số đã chọn
    label_indices_sets = {l: set(inverted_index.get(l, [])) for l in labels_in_order}
    achieved_counts = {}                     # nhãn -> số bản ghi thỏa chứa nhãn trong tập đã chọn

    for lbl in labels_in_order:
        lbl_set = label_indices_sets[lbl]
        already = len(selected & lbl_set)    # số dòng chứa nhãn này đã “có sẵn” trong tập
        missing = max(0, quota_per_label - already)

        if missing > 0:
            # ứng viên mới: các dòng chứa nhãn này nhưng chưa được chọn
            candidates = [i for i in inverted_index.get(lbl, []) if i not in selected]
            take = candidates[:missing]
            selected.update(take)

        # cập nhật lại achieved thực tế (sau khi thêm)
        achieved_counts[lbl] = len(selected & lbl_set)

    # Tạo df_out theo thứ tự xuất hiện ban đầu
    # (giữ nguyên thứ tự gốc để dễ kiểm tra)
    selected_sorted = sorted(list(selected))
    df_out = df.loc[selected_sorted].copy()

    return df_out, achieved_counts

# --------------------------
# 5) Pipeline chính
# --------------------------
def build_top50_quota_dataset(df_processed, categories_col="categories",
                              top_k=50, quota_per_label=300):
    # Chuẩn hoá
    df_norm = normalize_categories_col(df_processed, col=categories_col)

    # Inverted index
    inv = build_inverted_index(df_norm[categories_col])

    # Top-K nhãn con toàn bộ
    top_labels, global_counts = top_k_child_labels(inv, k=top_k)
    print(f"Top {top_k} nhãn con (theo số dòng chứa nhãn):")
    for i, lbl in enumerate(top_labels, 1):
        print(f"{i:>2}. {lbl}: {global_counts[lbl]} dòng")

    # Chọn theo quota
    print("\nĐang chọn dữ liệu theo quota mỗi nhãn =", quota_per_label)
    df_out, achieved = select_rows_by_label_quota(df_norm, inv, top_labels, quota_per_label=quota_per_label)

    # Thống kê đạt được
    print(f"\n✅ Hoàn tất. Số dòng trong tập kết quả: {len(df_out)}")
    print("Số dòng đạt được cho từng nhãn (đã tính cả 'ăn ké' từ nhãn trước):")
    for lbl in top_labels:
        got = achieved.get(lbl, 0)
        note = "" if got >= quota_per_label else f" (thiếu {quota_per_label - got})"
        print(f"- {lbl}: {got}{note}")

    return df_out, top_labels, achieved

# ============================
# CÁCH DÙNG:
# df_processed: DataFrame gốc, có cột 'categories'
# ============================

# df_selected, top50_labels, per_label_counts = build_top50_quota_dataset(
#     df_processed,
#     categories_col="categories",
#     top_k=50,
#     quota_per_label=300
# )

# # (tuỳ chọn) lưu ra CSV
# df_selected.to_csv("top50_child_labels_quota300.csv", index=False, encoding="utf-8")


In [99]:
df_selected, top50_labels, per_label_counts = build_top50_quota_dataset(
    df_processed,
    categories_col="categories",
    top_k=50,
    quota_per_label=300
)

# (tuỳ chọn) lưu ra CSV
df_selected.to_csv("top50_child_labels_quota300.csv", index=False, encoding="utf-8")


Top 50 nhãn con (theo số dòng chứa nhãn):
 1. hep-th: 11343 dòng
 2. hep-ph: 10527 dòng
 3. gr-qc: 8563 dòng
 4. quant-ph: 6814 dòng
 5. hep-ex: 4398 dòng
 6. nucl-th: 4221 dòng
 7. math-ph: 2792 dòng
 8. math.MP: 2792 dòng
 9. astro-ph.CO: 2425 dòng
10. hep-lat: 2155 dòng
11. nucl-ex: 2070 dòng
12. cs.LG: 2004 dòng
13. cond-mat.stat-mech: 1831 dòng
14. physics.optics: 1652 dòng
15. cond-mat.str-el: 1530 dòng
16. astro-ph.HE: 1468 dòng
17. cond-mat.mes-hall: 1453 dòng
18. astro-ph: 1352 dòng
19. stat.ML: 1266 dòng
20. stat.ME: 1100 dòng
21. q-fin.EC: 983 dòng
22. nlin.CD: 926 dòng
23. cs.CV: 899 dòng
24. econ.GN: 887 dòng
25. physics.atom-ph: 868 dòng
26. eess.SP: 826 dòng
27. stat.AP: 809 dòng
28. econ.EM: 760 dòng
29. nlin.SI: 759 dòng
30. cond-mat.mtrl-sci: 732 dòng
31. eess.IV: 644 dòng
32. q-bio.PE: 628 dòng
33. cs.SY: 624 dòng
34. math.PR: 612 dòng
35. nlin.PS: 584 dòng
36. eess.SY: 560 dòng
37. astro-ph.SR: 539 dòng
38. physics.soc-ph: 536 dòng
39. cs.AI: 533 dòng
40. cond-mat.q

# đếm số lần xuất hiện của df_selected

In [100]:
import pandas as pd
import ast
from collections import Counter

def normalize_categories(x):
    """Chuẩn hoá dữ liệu categories về list."""
    if isinstance(x, list):
        return x
    if pd.isna(x):
        return []
    if isinstance(x, str):
        try:
            v = ast.literal_eval(x)
            return v if isinstance(v, list) else [x.strip()]
        except Exception:
            return [x.strip()]
    return []

def count_label_frequencies(df: pd.DataFrame, col_name: str) -> pd.DataFrame:
    """
    Đếm số lần xuất hiện của tất cả nhãn trong một cột (multi-label).
    """
    all_labels = []
    for cats in df[col_name].apply(normalize_categories):
        all_labels.extend(cats)

    counts = Counter(all_labels)

    return (
        pd.DataFrame(counts.items(), columns=["nhãn", "số_lần_xuất_hiện"])
          .sort_values("số_lần_xuất_hiện", ascending=False)
          .reset_index(drop=True)
    )

df_child_counts = count_label_frequencies(df_selected, "categories")
print(df_child_counts)

         nhãn  số_lần_xuất_hiện
0    quant-ph              1457
1      hep-th               994
2      hep-ph               907
3       gr-qc               722
4       cs.LG               709
..        ...               ...
153     cs.PF                 1
154  chao-dyn                 1
155     cs.OH                 1
156   math.KT                 1
157   math.AC                 1

[158 rows x 2 columns]


In [101]:
df_child_counts

nhãn  số_lần_xuất_hiện
0    quant-ph              1457
1      hep-th               994
2      hep-ph               907
3       gr-qc               722
4       cs.LG               709
..        ...               ...
153     cs.PF                 1
154  chao-dyn                 1
155     cs.OH                 1
156   math.KT                 1
157   math.AC                 1

[158 rows x 2 columns]

# Qua từng dòng data bỏ nhãn phụ trong đó  và đếm số lần xuất hiện sau khi xử lý

In [85]:
list50labels = [
    "hep-th",
    "hep-ph",
    "gr-qc",
    "quant-ph",
    "hep-ex",
    "nucl-th",
    "math-ph",
    "math.MP",
    "astro-ph.CO",
    "hep-lat",
    "nucl-ex",
    "cs.LG",
    "cond-mat.stat-mech",
    "physics.optics",
    "cond-mat.str-el",
    "astro-ph.HE",
    "cond-mat.mes-hall",
    "astro-ph",
    "stat.ML",
    "stat.ME",
    "q-fin.EC",
    "nlin.CD",
    "cs.CV",
    "econ.GN",
    "physics.atom-ph",
    "eess.SP",
    "stat.AP",
    "econ.EM",
    "nlin.SI",
    "cond-mat.mtrl-sci",
    "eess.IV",
    "q-bio.PE",
    "cs.SY",
    "math.PR",
    "nlin.PS",
    "eess.SY",
    "astro-ph.SR",
    "physics.soc-ph",
    "cs.AI",
    "cond-mat.quant-gas",
    "astro-ph.GA",
    "q-bio.QM",
    "cond-mat.supr-con",
    "econ.TH",
    "q-fin.ST",
    "q-bio.NC",
    "physics.chem-ph",
    "math.OC",
    "math.AP",
    "cond-mat.other"
]


In [88]:
import ast

# chuẩn hoá categories thành list
def normalize_categories(x):
    if isinstance(x, list):
        return x
    if pd.isna(x):
        return []
    if isinstance(x, str):
        try:
            v = ast.literal_eval(x)
            return v if isinstance(v, list) else [x.strip()]
        except Exception:
            return [s.strip() for s in x.split(",")] if "," in x else [x.strip()]
    return []

# giả sử bạn có list 50 labels trong biến list50labels
# ví dụ: list50labels = ["hep-th", "hep-ph", "gr-qc", "quant-ph", ...]

# Tạo dataframe mới từ df_selected mà không tác động đến bản gốc
df_selected_top50 = df_selected.copy()
df_selected_top50["categories"] = df_selected_top50["categories"].apply(normalize_categories)

# lọc lại categories theo list50labels
df_selected_top50["categories"] = df_selected_top50["categories"].apply(
    lambda cats: [c for c in cats if c in list50labels]
)

# loại bỏ các dòng trống sau khi lọc
df_selected_top50 = df_selected_top50[df_selected_top50["categories"].map(len) > 0].reset_index(drop=True)

print("✅ Hoàn tất. Số dòng còn lại:", len(df_selected_top50))
print(df_selected_top50.head())


✅ Hoàn tất. Số dòng còn lại: 11240
                                             authors  \
0       Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1  David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2             Jiliang Jing, Qiyuan Pan, Songbai Chen   
3                Szabolcs Horv\'at and Carl D. Modes   
4                         A. Hopper and P. F. Barker   

                                               title  \
0  Power-law cosmological solution derived from D...   
1  PyFstat: a Python package for continuous gravi...   
2  Holographic Superconductors with Power-Maxwell...   
3  Connectedness matters: Construction and exact ...   
4  A hybrid quantum system formed by trapping ato...   

                                            abstract  \
0    By studying a tachyon field on the DGP brane...   
1    Gravitational waves in the sensitivity band ...   
2    With the Sturm-Liouville analytical and nume...   
3    We describe a new method for the random samp...   
4    Near-f

In [89]:
df_child_counts = count_label_frequencies(df_selected, "categories")
print(df_child_counts)

                  nhãn  số_lần_xuất_hiện
0             quant-ph              1457
1               hep-th               994
2               hep-ph               907
3                gr-qc               722
4                cs.LG               709
5              math-ph               626
6              math.MP               626
7   cond-mat.stat-mech               447
8       physics.optics               419
9              stat.ML               410
10     cond-mat.str-el               405
11   cond-mat.mes-hall               390
12         astro-ph.CO               383
13               cs.CV               380
14             stat.ME               369
15         astro-ph.HE               355
16            q-fin.EC               354
17              hep-ex               351
18               cs.SY               343
19     physics.atom-ph               343
20             nucl-th               341
21             nlin.CD               340
22             eess.SP               336
23             s

In [91]:
output_file = "/content/drive/MyDrive/AIO/df_selected_top50.csv"
df_selected_top50.to_csv(output_file, index=False, encoding="utf-8")

print(f"💾 Đã lưu DataFrame vào file: {output_file}")

💾 Đã lưu DataFrame vào file: /content/drive/MyDrive/AIO/df_selected_top50.csv


### ANN - Tính toán cha con

In [111]:
df_selected_top50

authors  \
0           Yongli Ping, Lixin Xu, Hongya Liu, Ying Shao   
1      David Keitel, Rodrigo Tenorio, Gregory Ashton,...   
2                 Jiliang Jing, Qiyuan Pan, Songbai Chen   
3                    Szabolcs Horv\'at and Carl D. Modes   
4                             A. Hopper and P. F. Barker   
...                                                  ...   
11235  A. A. Banishev, J. Wagner, T. Emig, R. Zandi, ...   
11236  A. A. Demenev, A. A. Shchekin, A. V. Larionov,...   
11237     Stefan Rist, Patrizia Vignolo, Giovanna Morigi   
11238  Meg Mahat (1), Sergio Periera (2), Ian M. Wats...   
11239   Yaron Bromberg, Yoav Lahini and Yaron Silberberg   

                                                   title  \
0      Power-law cosmological solution derived from D...   
1      PyFstat: a Python package for continuous gravi...   
2      Holographic Superconductors with Power-Maxwell...   
3      Connectedness matters: Construction and exact ...   
4      A hybrid quantum system formed by trapping ato...   
...                                                  ...   
11235  Demonstration of Angle Dependent Casimir Force...   
11236  Kinetics of stimulated polariton scattering in...   
11237  Photonic spectrum of bichromatic optical lattices   
11238  Coherent Control of Acoustic Phonons in hybrid...   
11239       Bloch oscillations of Path-Entangled Photons   

                                                abstract  \
0        By studying a tachyon field on the DGP brane...   
1        Gravitational waves in the sensitivity band ...   
2        With the Sturm-Liouville analytical and nume...   
3        We describe a new method for the random samp...   
4        Near-field, radially symmetric optical poten...   
...                                                  ...   
11235    The normal Casimir force between a sinusoida...   
11236    We demonstrate for the first time the strong...   
11237    We study the photonic spectrum of a one-dime...   
11238    We demonstrate the modification of coherent ...   
11239    We show that when photons in N-particle path...   

                                        categories  \
0                                         [hep-th]   
1                                          [gr-qc]   
2                                  [hep-th, gr-qc]   
3                                 [physics.soc-ph]   
4      [quant-ph, physics.atom-ph, physics.optics]   
...                                            ...   
11235                   [quant-ph, cond-mat.other]   
11236                             [cond-mat.other]   
11237                   [quant-ph, cond-mat.other]   
11238          [cond-mat.mtrl-sci, cond-mat.other]   
11239                   [quant-ph, cond-mat.other]   

                                         json_categories  
0                                 {"hep-th": ["hep-th"]}  
1      {"gr-qc": ["gr-qc"], "astro-ph": ["astro-ph.IM"]}  
2             {"hep-th": ["hep-th"], "gr-qc": ["gr-qc"]}  
3                        {"physics": ["physics.soc-ph"]}  
4      {"quant-ph": ["quant-ph"], "physics": ["physic...  
...                                                  ...  
11235  {"quant-ph": ["quant-ph"], "cond-mat": ["cond-...  
11236                   {"cond-mat": ["cond-mat.other"]}  
11237  {"quant-ph": ["quant-ph"], "cond-mat": ["cond-...  
11238  {"cond-mat": ["cond-mat.mtrl-sci", "cond-mat.o...  
11239  {"quant-ph": ["quant-ph"], "cond-mat": ["cond-...  

[11240 rows x 5 columns]

In [105]:
!pip install pandas numpy scikit-learn sentence-transformers
!pip install faiss-gpu  # nếu có GPU CUDA


ERROR: Could not find a version that satisfies the requirement faiss-gpu (from versions: none)
ERROR: No matching distribution found for faiss-gpu


In [107]:
!pip install faiss-cpu

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.4/31.4 MB 58.4 MB/s eta 0:00:00


In [113]:
import pandas as pd
import numpy as np
import ast
import time

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.metrics import f1_score, accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sentence_transformers import SentenceTransformer

# ----------------- 1. LOAD DATA -----------------
df = df_selected_top50.copy()

# Ghép 3 cột authors + title + abstract thành text
df["text"] = (
    df["authors"].astype(str) + " " +
    df["title"].astype(str) + " " +
    df["abstract"].astype(str)
)

texts = df["text"].tolist()
labels = df["categories"].tolist()
print(f"✅ Data: {len(df)} samples | text + categories ready")

# ----------------- 2. LABELS -----------------
mlb = MultiLabelBinarizer()
Y = mlb.fit_transform(labels)
print(f"✅ {len(mlb.classes_)} unique labels. Ví dụ: {mlb.classes_[:10]}")

# ----------------- 3. EMBEDDINGS -----------------
print("🔄 Đang tạo embedding với SentenceTransformer...")
start = time.time()
model = SentenceTransformer("all-MiniLM-L6-v2", device="cuda")
X = model.encode(texts, show_progress_bar=True, device="cuda")
X = np.array(X, dtype="float32")
print(f"✅ Embedding xong: {X.shape} (thời gian {time.time()-start:.2f}s)")




✅ Data: 11240 samples | text + categories ready
✅ 50 unique labels. Ví dụ: ['astro-ph' 'astro-ph.CO' 'astro-ph.GA' 'astro-ph.HE' 'astro-ph.SR'
 'cond-mat.mes-hall' 'cond-mat.mtrl-sci' 'cond-mat.other'
 'cond-mat.quant-gas' 'cond-mat.stat-mech']
🔄 Đang tạo embedding với SentenceTransformer...


Batches:   0%|          | 0/352 [00:00<?, ?it/s]

✅ Embedding xong: (11240, 384) (thời gian 30.79s)
✅ Split: train=8992, test=2248
🔄 Huấn luyện KNN...
✅ Train xong (thời gian 0.06s)
🎯 Đánh giá KNN → Accuracy=0.3394, F1-micro=0.6141, F1-macro=0.5676
📊 Ví dụ 5 sample đầu tiên trong predict_df:
                                                 authors  \
4170   Yiju Ma, Kevin Swandi, Archie Chapman and Greg...   
10819      Ben Roig, Michael R. Blanton, Nicolas P. Ross   
318    Mustapha Azreg-A\"inou, Glauber T. Marques, Ma...   
3473   J. Billy (LCFIO), V. Josse (LCFIO), Z. Zuo (LC...   
9959   Thanh Nguyen, Vy Bui, Anh Thai, Van Lam, Chris...   

                                                   title  \
4170   Multi-Stage Compound Real Options Valuation in...   
10819  Unusual Broad-Line MgII Emitters Among Luminou...   
318           Phantom black holes and critical phenomena   
3473   Guided atom laser : a new tool for guided atom...   
9959   Virtual organelle self-coding for fluorescence...   

                      true_labels  

In [125]:
!pip install iterative-stratification

In [127]:

from sklearn.metrics import classification_report, jaccard_score
from iterstrat.ml_stratifiers import MultilabelStratifiedKFold, MultilabelStratifiedShuffleSplit

# ----------------- 4. TRAIN/TEST SPLIT (MULTILABEL STRATIFY) -----------------
msss = MultilabelStratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
for train_idx, test_idx in msss.split(X, Y):
    X_train, X_test = X[train_idx], X[test_idx]
    y_train, y_test = Y[train_idx], Y[test_idx]
    idx_train, idx_test = df.index[train_idx], df.index[test_idx]

print(f"✅ Split: train={X_train.shape[0]}, test={X_test.shape[0]}")

# ----------------- 5. TRAIN KNN -----------------
print("🔄 Huấn luyện KNN...")
knn = KNeighborsClassifier(n_neighbors=5, metric="cosine")
start = time.time()
knn.fit(X_train, y_train)
print(f"✅ Train xong (thời gian {time.time()-start:.2f}s)")

# ----------------- 6. DỰ ĐOÁN -----------------
y_pred = knn.predict(X_test)
acc = accuracy_score(y_test, y_pred)
f1_micro = f1_score(y_test, y_pred, average="micro")
f1_macro = f1_score(y_test, y_pred, average="macro")
jaccard = jaccard_score(y_test, y_pred, average="samples")  # dùng trung bình theo từng sample

print(f"🎯 Đánh giá KNN → Accuracy={acc:.4f}, F1-micro={f1_micro:.4f}, "
      f"F1-macro={f1_macro:.4f}, Jaccard={jaccard:.4f}")

# ----------------- 7. TẠO predict_df -----------------
true_labels = mlb.inverse_transform(y_test)
pred_labels = mlb.inverse_transform(y_pred)

predict_df = df.loc[idx_test].copy()
predict_df["true_labels"] = true_labels
predict_df["predicted_labels"] = pred_labels

print("📊 Ví dụ 5 sample đầu tiên trong predict_df:")
print(predict_df[["authors", "title", "true_labels", "predicted_labels"]].head())

# ----------------- 8. SAVE KẾT QUẢ -----------------
predict_df.to_csv("/content/drive/MyDrive/AIO/predict_results.csv", index=False, encoding="utf-8")
print("💾 Đã lưu kết quả dự đoán vào file: /content/drive/MyDrive/AIO/predict_results.csv")

# ----------------- 9. REPORT CHI TIẾT -----------------
print("\n📊 Classification Report chi tiết cho từng nhãn:")
report = classification_report(y_test, y_pred, target_names=mlb.classes_, zero_division=0)
print(report)


✅ Split: train=8991, test=2249
🔄 Huấn luyện KNN...
✅ Train xong (thời gian 0.05s)
🎯 Đánh giá KNN → Accuracy=0.3433, F1-micro=0.6147, F1-macro=0.5653, Jaccard=0.5227
📊 Ví dụ 5 sample đầu tiên trong predict_df:
                                              authors  \
2              Jiliang Jing, Qiyuan Pan, Songbai Chen   
4                          A. Hopper and P. F. Barker   
6   Christian Majenz, Maris Ozols, Christian Schaf...   
9                  Alexander Grant, Eanna E. Flanagan   
11                                   Kaushik Majumdar   

                                                title  \
2   Holographic Superconductors with Power-Maxwell...   
4   A hybrid quantum system formed by trapping ato...   
6             Local simultaneous state discrimination   
9   Non-conservation of Carter in black hole space...   
11  Fourier Uniformity: An Useful Tool for Analyzi...   

                                    true_labels  \
2                               (gr-qc, hep-th)   
4  

Rồi 👍 mình viết lại một bản **report gọn – rõ – dễ hiểu** cho kết quả của bạn.

---

# 📑 Report Kết Quả Mô Hình KNN + ANN

## 1. Ý nghĩa các metrics

### 🔹 Accuracy

* Trong **multi-label classification**, một sample chỉ được coi là đúng khi **toàn bộ tập nhãn** khớp với ground truth.
* Do đó Accuracy thường thấp hơn so với F1.

### 🔹 F1-micro

* Tính **TP, FP, FN trên toàn bộ dataset** (gộp hết các nhãn).
* **Ưu tiên nhãn phổ biến** (nhiều dữ liệu).
* Đo mức độ mô hình làm tốt trên toàn cục.

### 🔹 F1-macro

* Tính **F1 riêng cho từng nhãn**, sau đó lấy trung bình cộng.
* **Tất cả nhãn được coi trọng như nhau** (cân bằng giữa nhãn nhiều và nhãn ít).
* Đo mức độ mô hình có **công bằng với các nhãn nhỏ/lẻ** hay không.

---

## 2. Phân tích kết quả

* **Accuracy = 0.3394 (~34%)**
  → Nghĩa là chỉ 34% sample có toàn bộ nhãn dự đoán trùng khớp hoàn toàn với nhãn thật. Đây là con số **khá ổn** trong bối cảnh multi-label, vì chỉ cần sai 1 nhãn cũng bị tính là sai cả sample.

* **F1-micro = 0.6141 (~61%)**
  → Khi xét toàn bộ nhãn (ưu tiên nhãn phổ biến như *hep-th, hep-ph, gr-qc*), mô hình dự đoán **khá tốt**.

* **F1-macro = 0.5676 (~57%)**
  → Khi tính công bằng cho từng nhãn (bao gồm cả nhãn ít dữ liệu như *econ.TH, math.OC, q-bio.*), mô hình **yếu hơn rõ rệt**.
  → Điều này chứng tỏ:

  * Model **mạnh với nhãn phổ biến**.
  * **Yếu với nhãn hiếm**, dẫn tới mất cân bằng.


```
🎯 Đánh giá KNN → Accuracy=0.3394, F1-micro=0.6141, F1-macro=0.5676, Jaccard=0.5195

💾 Đã lưu kết quả dự đoán vào file: /content/drive/MyDrive/AIO/predict_results.csv

📊 Classification Report chi tiết cho từng nhãn:
                    precision    recall  f1-score   support

          astro-ph       0.47      0.12      0.19        59
       astro-ph.CO       0.57      0.41      0.48        73
       astro-ph.GA       0.53      0.47      0.50        58
       astro-ph.HE       0.69      0.50      0.58        68
       astro-ph.SR       0.75      0.44      0.56        54
 cond-mat.mes-hall       0.63      0.31      0.42        86
 cond-mat.mtrl-sci       0.92      0.56      0.69        61
    cond-mat.other       0.18      0.05      0.07        66
cond-mat.quant-gas       0.53      0.37      0.44        65
cond-mat.stat-mech       0.41      0.13      0.20       101
   cond-mat.str-el       0.49      0.45      0.46        74
 cond-mat.supr-con       0.73      0.67      0.70        52
             cs.AI       0.67      0.23      0.34        70
             cs.CV       0.78      0.79      0.79        73
             cs.LG       0.62      0.62      0.62       128
             cs.SY       0.80      0.61      0.69        66
           econ.EM       0.75      0.51      0.61        59
           econ.GN       0.52      0.30      0.38        53
           econ.TH       0.74      0.76      0.75        59
           eess.IV       0.80      0.82      0.81        60
           eess.SP       0.85      0.59      0.70        68
           eess.SY       0.77      0.60      0.67        60
             gr-qc       0.88      0.72      0.79       147
            hep-ex       0.83      0.62      0.71        77
           hep-lat       0.81      0.81      0.81        62
            hep-ph       0.82      0.73      0.77       184
            hep-th       0.79      0.63      0.70       215
           math-ph       0.69      0.46      0.55       135
           math.AP       0.79      0.46      0.58        65
           math.MP       0.69      0.46      0.55       135
           math.OC       0.48      0.30      0.37        50
           math.PR       0.89      0.48      0.63        66
           nlin.CD       0.85      0.52      0.64        64
           nlin.PS       0.67      0.49      0.57        57
           nlin.SI       0.67      0.66      0.67        65
           nucl-ex       0.85      0.84      0.85        62
           nucl-th       0.69      0.52      0.59        64
   physics.atom-ph       0.73      0.50      0.59        70
   physics.chem-ph       0.71      0.45      0.55        67
    physics.optics       0.61      0.43      0.50        82
    physics.soc-ph       0.56      0.30      0.39        67
          q-bio.NC       0.93      0.80      0.86        64
          q-bio.PE       0.81      0.81      0.81        57
          q-bio.QM       0.50      0.28      0.36        54
          q-fin.EC       0.59      0.36      0.45        61
          q-fin.ST       0.75      0.65      0.69        65
          quant-ph       0.86      0.83      0.84       320
           stat.AP       0.44      0.09      0.15        77
           stat.ME       0.51      0.32      0.39        69
           stat.ML       0.43      0.35      0.38        72

         micro avg       0.72      0.53      0.61      4086
         macro avg       0.68      0.50      0.57      4086
      weighted avg       0.70      0.53      0.59      4086
       samples avg       0.65      0.57      0.58      4086

```

```bash
                  nhãn  số_lần_xuất_hiện
0             quant-ph              1457
1               hep-th               994
2               hep-ph               907
3                gr-qc               722
4                cs.LG               709
5              math-ph               626
6              math.MP               626
7   cond-mat.stat-mech               447
8       physics.optics               419
9              stat.ML               410
10     cond-mat.str-el               405
11   cond-mat.mes-hall               390
12         astro-ph.CO               383
13               cs.CV               380
14             stat.ME               369
15         astro-ph.HE               355
16            q-fin.EC               354
17              hep-ex               351
18               cs.SY               343
19     physics.atom-ph               343
20             nucl-th               341
21             nlin.CD               340
22             eess.SP               336
23             stat.AP               331
24   cond-mat.mtrl-sci               327
25             nlin.SI               325
26             econ.GN               321
27             hep-lat               316
28             nlin.PS               314
29             econ.EM               313
30            q-bio.PE               308
31            astro-ph               307
32             eess.IV               306
33             nucl-ex               306
34             math.PR               306
35             eess.SY               306
36      physics.soc-ph               305
37  cond-mat.quant-gas               303
38               cs.AI               302
39             econ.TH               302
40     physics.chem-ph               301
41            q-bio.NC               300
42            q-bio.QM               300
43   cond-mat.supr-con               300
44             math.OC               300
45            q-fin.ST               300
46      cond-mat.other               300
47         astro-ph.SR               300
48         astro-ph.GA               300
49             math.AP               300
```

# Model embedding mạnh hơn